### PR TITLE
fix: derive decimal rounding result metadata

### LINKED ISSUE #28593: [Compatibility]: TRUNCATE retains DECIMAL input scale instead of the requested precision
### Is there an existing issue for the same bug?

- [x] I have checked the existing issues, including open and closed issues.

### Branch Name

main

### Commit ID

fc621e3616d229c7a29d0c80e73ed8eb1997459c

### Other Environment Information

- MatrixOne: locally built from the commit above (`8.0.30-MatrixOne-v1.3.0`)
- MySQL baseline: 8.0.46
- Verification date: 2026-09-10

### Actual Behavior

For exact DECIMAL input and a nonnegative constant precision, MatrixOne's `TRUNCATE(X,D)` always retains the input DECIMAL precision and scale. MySQL reduces the result scale to `min(input_scale,D)` and removes the discarded fractional capacity from the precision.

This is visible in both returned text and persisted schema. For a `DECIMAL(10,4)` source:

```text
expression       MatrixOne value/type    MySQL value/type
TRUNCATE(d,2)    12345.6700 DECIMAL(10,4) 12345.67 DECIMAL(8,2)
TRUNCATE(d,0)    12345.0000 DECIMAL(10,4) 12345    DECIMAL(6,0)
TRUNCATE(d,8)    12345.6789 DECIMAL(10,4) 12345.6789 DECIMAL(10,4)
```

The same mismatch spans MatrixOne's three exact-decimal representations:

- `DECIMAL(10,4)` truncated to 2 remains `DECIMAL(10,4)` instead of `DECIMAL(8,2)`.
- `DECIMAL(38,10)` truncated to 5 remains `DECIMAL(38,10)` instead of `DECIMAL(33,5)`.
- `DECIMAL(50,10)` truncated to 5 remains `DECIMAL(50,10)` instead of `DECIMAL(45,5)`.

`CREATE TABLE ... AS SELECT` persists those wider MatrixOne types, and a view exposes the same wider metadata for nonnegative `D`. This can change downstream client metadata, generated schemas, storage width, string output, and comparisons that depend on implicit type derivation even though the significant numeric digits agree.

The negative-`D` case is deliberately not used as the compatibility contract here because MySQL 8.0.46 exposes different metadata for that case through a view versus CTAS. The clear mismatch above uses only nonnegative constant precision, for which MySQL is consistent.

### Expected Behavior

For exact DECIMAL input and nonnegative constant `D`, `TRUNCATE(X,D)` should derive a MySQL-compatible result type:

- result scale = `min(input scale, D)`;
- integer-digit capacity is preserved;
- no extra fractional zeroes remain after the requested truncation;
- CTAS and view metadata reflect the derived type.

### Steps to Reproduce

```sql
DROP DATABASE IF EXISTS truncate_decimal_type;
CREATE DATABASE truncate_decimal_type;
USE truncate_decimal_type;

CREATE TABLE src(
  id INT PRIMARY KEY,
  d64 DECIMAL(10,4),
  d128 DECIMAL(38,10),
  d256 DECIMAL(50,10)
);
INSERT INTO src VALUES
  (1,12345.6789,
     1234567890123456789012345678.1234567890,
     1234567890123456789012345678901234567890.1234567890),
  (2,-12345.6789,-1.1234567890,-1.1234567890);

SELECT id,
       CAST(TRUNCATE(d64,2) AS CHAR),
       CAST(TRUNCATE(d64,0) AS CHAR),
       CAST(TRUNCATE(d64,8) AS CHAR),
       CAST(TRUNCATE(d128,5) AS CHAR),
       CAST(TRUNCATE(d256,5) AS CHAR)
FROM src ORDER BY id;

CREATE TABLE dst AS
SELECT TRUNCATE(d64,2) AS d2,
       TRUNCATE(d64,0) AS d0,
       TRUNCATE(d64,8) AS d8,
       TRUNCATE(d128,5) AS d128_5,
       TRUNCATE(d256,5) AS d256_5
FROM src;

SHOW CREATE TABLE dst;
SELECT column_name,column_type,numeric_precision,numeric_scale
FROM information_schema.columns
WHERE table_schema=DATABASE() AND table_name='dst'
ORDER BY ordinal_position;
```

### Additional information

The DECIMAL overloads for `TRUNCATE` in `pkg/sql/plan/function/list_builtIn.go` return `parameters[0]` unchanged for Decimal64, Decimal128, and Decimal256. The evaluator in `func_binary.go` can rescale when the result vector's scale differs, but the declared return type never supplies the constant `D`-derived scale, so it retains the input scale and appends zeroes.

Positive, negative, zero, `D=0`, `0<D<input_scale`, `D>input_scale`, Decimal64/128/256, direct SELECT, CTAS, view, and information-schema paths were reproduced three times on the same latest-main build. Closed #28492 concerned rejection of dynamic or NULL precision arguments; this report concerns the result type and representation of accepted constant-precision calls.

### PR DESCRIPTION
**Pull request by daviszhen**
## What type of PR is this?

- [ ] API-change
- [x] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

Fixes #28593
Part of #28595

## What this PR does / why we need it:

- Derives ROUND and TRUNCATE precision and scale from a constant digits argument.
- Refines CEIL, CEILING, and FLOOR decimal precision/scale for both one-argument and constant two-argument forms.
- Keeps the existing Decimal64/Decimal128/Decimal256 physical result type so serialized plans remain executable by older CNs during rolling upgrades; BIGINT lowering is deferred until it can be protected by a protocol-version gate.
- Preserves DECIMAL(19,0) values outside the signed BIGINT domain without wrapping.
- Adds binder and execution regression coverage around MaxInt64/MinInt64 and the full DECIMAL(19,0) boundary.

## Testing

- `CGO_CFLAGS="-I$PWD/thirdparties/install/include" CGO_LDFLAGS="-L$PWD/thirdparties/install/lib" go test ./pkg/sql/plan/function ./pkg/sql/plan -count=1`
- `make build`

### CODE REVIEW

**Comment by daviszhen** [comment]
已更新到最新 upstream/main（3232a9df2c），并修复上一轮唯一的 SCA 错误，当前 head 为 0b57ffef56。\n\n修复内容：删除 func_binary_test.go 中 Go 1.22+ 已不需要的 function := function 循环变量副本，消除 copyloopvar 告警。上一轮 UT、Coverage 和全部 BVT 均已通过。\n\n本地验证：\n- CI 同版本 golangci-lint v2.6.2 定向检查 ./pkg/sql/plan/function：0 issues\n- go test ./pkg/sql/plan ./pkg/sql/plan/function -count=1：通过\n- make build：通过\n- git diff --check：通过

**Comment by daviszhen** [comment]
补充：验证期间 main 又合入两个提交，现已再次同步到当前最新 upstream/main 9a5dd0c6e3；最终 head 为 00c81bf840，更新后 make build 再次通过。

**Comment by daviszhen** [comment]
Updated to current upstream/main (cd8e8d9134) in e98c59b344.\n\nThe prior Ubuntu/x86 failure was the overall UT deadline (exit 124) while the embedded-cluster stage was still running; no test assertion failed. The branch now includes main's bounded UT overlap and timeout scheduling update (#28643).\n\nVerification:\n- make build\n- go test ./pkg/sql/plan/function ./pkg/sql/plan -count=1\n- golangci-lint v2.6.2 on affected packages: 0 issues\n\nNew CI: https://github.com/matrixorigin/matrixone/actions/runs/34557993865

File: pkg/sql/plan/base_binder.go [hunk 1]
```diff
@@ -6017,6 +5997,75 @@ func bindFuncExprImplByPlanExpr(
...
+			return
+		}
+		precision := integerDigits
+		if input.Scale > 0 {
+			precision++ // a fractional value can carry into a new integer digit
+		}
+		if precision < 1 {
+			precision = 1
+		}
+		if precision <= 19 {
```
Comment (XuPeng-SH) [inline pkg/sql/plan/base_binder.go]: [P2] Do not narrow the full DECIMAL(19,0) domain to signed BIGINT

Decimal precision 19 does not imply signed-int64 representability. For a DECIMAL(19,0) input this condition selects T_int64, and the new CeilDecimal128/FloorDecimal128 branches return int64(result.B0_63) without a range check. For example:

```sql
CREATE TABLE t (d DECIMAL(19,0));
INSERT INTO t VALUES (9999999999999999999);
SELECT d, CEIL(d), FLOOR(d) FROM t;
```

Both functions must preserve this already-integral value. The new path instead returns -8446744073709551617 (9999999999999999999 interpreted as signed 64-bit), silently changing sign. This follows directly from Decimal128.Ceil/Floor returning x when both scales are zero, followed by the new unchecked conversion. It affects column execution and consequently CTAS/result metadata, not only constant folding.

Choose BIGINT only when the full input/result domain fits it; keep a decimal result for the unsafe 19-integer-digit domain (or implement an explicitly compatible checked conversion, never wrapping). Please add positive and negative boundaries around MaxInt64/MinInt64, including DECIMAL(19,0), rather than only small representative values.

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
REQUEST_CHANGES at e98c59b344e2cd7fce8b4d6eef816c9e40dde335 against base cd8e8d91343798f565f8ad0f8473cb995b8b361b: one concrete numeric-result blocker, detailed inline. The CEIL/FLOOR BIGINT admission rule admits DECIMAL(19,0) values outside signed int64 and the executor silently wraps them.

Reviewed the full four-file change, current comments, binder refinement and overload return-type callbacks, and Decimal128 rounding/conversion implementations. Nonnegative ROUND/TRUNCATE precision reduction reserves the ROUND carry digit and retains the physical decimal family; dynamic/NULL digits keep the prior conservative path. The new tests cover metadata derivation and ordinary positive/negative execution across decimal widths, but do not cover the failing BIGINT boundary.

Evidence is exact-head source/control-flow analysis plus independent integer arithmetic for the counterexample. I did not run native UT/BVT or a live SQL server, and did not wait for CI; author-reported package/build/CTAS checks are separate evidence. This is a concrete result regression, not a blocker solely for optional testing. No additional blocker established; mixed-version execution of the changed physical result type remains an implementation validation concern, not certified by these unit tests. Head/base were rechecked unchanged immediately before submission.

File: pkg/sql/plan/base_binder.go (hunk 1 above)
Comment (aptend) [inline pkg/sql/plan/base_binder.go]: [P1] Keep `DECIMAL(19,0)` out of the int64 path. `precision <= 19` is not sufficient: signed BIGINT only covers up to 9,223,372,036,854,775,807, while `DECIMAL(19,0)` permits 9,999,999,999,999,999,999. `CEIL(CAST('9999999999999999999' AS DECIMAL(19,0)))` is already integral, but the new unchecked `B0_63` cast returns `-8446744073709551617` instead of the decimal value. I reproduced this on this head with a binder boundary test (the plan type is BIGINT). Only use BIGINT when the full declared domain is signed-int64-safe; otherwise preserve decimal.

File: pkg/sql/plan/base_binder.go [hunk 2]
```diff
@@ -6017,6 +5997,75 @@ func bindFuncExprImplByPlanExpr(
...
+func refineDecimalRoundingReturnType(name string, args []*plan.Expr, argsType []types.Type, returnType *types.Type) {
+	if len(argsType) == 0 || !argsType[0].IsDecimal() {
+		return
+	}
+	input := argsType[0]
+	integerDigits := input.Width - input.Scale
+
+	switch name {
+	case "ceil", "ceiling", "floor":
+		if len(args) != 1 {
```
Comment (aptend) [inline pkg/sql/plan/base_binder.go]: [P1] Preserve metadata refinement for the existing two-argument CEIL/FLOOR overloads. Returning here drops the behavior replaced above: for `d DECIMAL(10,4)`, `CEIL(d,2)` previously bound as `DECIMAL(10,2)`; this head binds `DECIMAL(10,4)`. The evaluator then sees unchanged result scale and returns/persists `123.4600` for `123.4567` instead of metadata/result `123.46` (for example, CTAS gets scale 4). I reproduced the exact binder types in a temporary test. Please derive the literal-digits case for CEIL/CEILING/FLOOR as before, with the corrected width rules.

File: pkg/sql/plan/function/func_binary.go [hunk 3]
```diff
@@ -289,6 +289,11 @@ func ceilDecimal128(x types.Decimal128, digits int64, scale int32, isConst bool)
 
 func CeilDecimal64(ivecs []*vector.Vector, result vector.FunctionResultWrapper, proc *process.Process, length int, selectList *FunctionSelectList) (err error) {
 	scale := ivecs[0].GetType().Scale
```
Comment (aptend) [inline pkg/sql/plan/function/func_binary.go:291]: [P1] Gate the new physical result type for rolling upgrades. The result wrapper is constructed from the serialized plan type. An upgraded coordinator can plan single-argument decimal CEIL as int64, but an old CN's `CeilDecimal64/128/256` lacks this branch and falls through to `generalMathMulti`; its `opUnaryFixedToFixed[T,T]` calls `MustFunctionResult[Decimal*]` on the int64 wrapper and panics with `wrong type for FunctionResultWrapper`. Thus a query pushed from a new coordinator to an old CN fails during a rolling upgrade. This PR has no MORPC version bump/gate. Please keep old-compatible lowering until remote CNs advertise support, or add a coordinated protocol gate plus a mixed-version test.

**Review by aptend — CHANGES_REQUESTED** [review]
Deep-reviewed exact head `e98c59b344e2cd7fce8b4d6eef816c9e40dde335` against base `cd8e8d91343798f565f8ad0f8473cb995b8b361b`, including the full diff, all existing reviews/comments, and both linked issues. The intended same-version metadata tests pass, but three blocking correctness/compatibility cases remain. I reproduced the first two with temporary boundary tests; those review-only tests were removed afterward. Exact-head CI is green.

File: pkg/sql/plan/base_binder.go (hunk 1 above)
Comment (daviszhen) [reply pkg/sql/plan/base_binder.go]: Fixed in c8d4d6d0a1. CEIL/FLOOR now preserve the Decimal128 physical result for DECIMAL(19,0), and execution tests cover MaxInt64, MaxInt64+1, MinInt64, MinInt64-1, and both 19-digit extrema without wrapping.

File: pkg/sql/plan/function/func_binary.go (hunk 3 above)
Comment (daviszhen) [reply pkg/sql/plan/function/func_binary.go:291]: Fixed in c8d4d6d0a1. The new int64 executor paths were removed and the binder keeps the original Decimal64/Decimal128/Decimal256 OID while refining only width/scale, so serialized plans remain executable by older CNs during rolling upgrades.

File: pkg/sql/plan/base_binder.go (hunk 2 above)
Comment (daviszhen) [reply pkg/sql/plan/base_binder.go]: Fixed in c8d4d6d0a1. Constant two-argument CEIL/CEILING/FLOOR now refine both precision and scale; DECIMAL(10,4) with digits=2 binds as DECIMAL(9,2), reserving one carry digit, and all three spellings have binder regression coverage.

File: pkg/sql/plan/base_binder.go (hunk 1 above)
Comment (daviszhen) [reply pkg/sql/plan/base_binder.go]: Fixed in c8d4d6d0a1. The BIGINT lowering was removed entirely, so the full DECIMAL(19,0) domain stays Decimal128 and no unchecked B0_63 conversion remains.

**Comment by daviszhen** [comment]
Addressed all current reviewer comments in  (fix commit ):\n\n- removed Decimal-to-BIGINT physical lowering so plans remain compatible with older CNs during rolling upgrades;\n- kept  in Decimal128 and added execution coverage at MaxInt64/MinInt64, adjacent out-of-range values, and both 19-digit extrema;\n- restored constant two-argument // metadata refinement, including carry-aware precision;\n- merged the latest .\n\nLocal verification passed:\n\n- FAIL	github.com/matrixorigin/matrixone/pkg/sql/plan/function [build failed]
FAIL	github.com/matrixorigin/matrixone/pkg/sql/plan [build failed]
FAIL\n- [Verify Go package graph]
mkdir -p install
mkdir -p install/include install/lib
using verified installed ONNX Runtime 1.26.0 from install/lib/onnxruntime_arm64.dylib
make[1]: Nothing to be done for `all'.
[Stage jieba dict]
rm -rf /Users/pengzhen/Documents/GitHub/matrixone-28593-rounding/dict
cp -r /Users/pengzhen/Documents/GitHub/matrixone-28593-rounding/pkg/monlp/tokenizer/dict /Users/pengzhen/Documents/GitHub/matrixone-28593-rounding/dict
[Build binary]
CGO_CFLAGS="-I/Users/pengzhen/Documents/GitHub/matrixone-28593-rounding/cgo -I/Users/pengzhen/Documents/GitHub/matrixone-28593-rounding/thirdparties/install/include " go build -mod=readonly   -ldflags="-extldflags '-L/Users/pengzhen/Documents/GitHub/matrixone-28593-rounding/cgo -lmo -L/Users/pengzhen/Documents/GitHub/matrixone-28593-rounding/thirdparties/install/lib -Wl,-rpath,@executable_path/lib' -X 'github.com/matrixorigin/matrixone/pkg/version.GoVersion=go version go1.26.5 darwin/arm64' -X 'github.com/matrixorigin/matrixone/pkg/version.BranchName=codex/decimal-rounding-metadata' -X 'github.com/matrixorigin/matrixone/pkg/version.CommitID=f337f93a30' -X 'github.com/matrixorigin/matrixone/pkg/version.BuildTime=1789110030' -X 'github.com/matrixorigin/matrixone/pkg/version.Version=codex/decimal-rounding-metadata'"   -o mo-service ./cmd/mo-service\n\nAll four review threads have been replied to and resolved. CI for the new head is running.

**Comment by daviszhen** [comment]
Addressed all current reviewer comments in `f337f93a30` (fix commit `c8d4d6d0a1`):

- removed Decimal-to-BIGINT physical lowering so plans remain compatible with older CNs during rolling upgrades;
- kept `DECIMAL(19,0)` in Decimal128 and added execution coverage at MaxInt64/MinInt64, adjacent out-of-range values, and both 19-digit extrema;
- restored constant two-argument `CEIL`/`CEILING`/`FLOOR` metadata refinement, including carry-aware precision;
- merged the latest `main`.

Local verification passed:

- `go test ./pkg/sql/plan/function ./pkg/sql/plan -count=1`
- `make build`

All four review threads have been replied to and resolved. CI for the new head is running.

**Comment by daviszhen** [comment]
Resolved the failing coverage checks at head `77537dea5b`.

Root cause from the coverage diagnostics: the only failed assertion was `pkg/logservice/TestProbeTestPortRejectsOccupiedSockets/udp`. The test selected an ephemeral port using UDP and then assumed that the same numeric TCP port was free. TCP and UDP allocate ephemeral ports independently, so CI already had that TCP port occupied. The later aggregate Coverage failure was only a consequence of the missing successful UT coverage profile.

Fix:

- select and temporarily reserve the ephemeral port with TCP first;
- bind the UDP fixture on that exact port before releasing TCP;
- retain the original assertion that a failed UDP probe closes its temporary TCP listener;
- merge current upstream `main` through `b349fea56f`.

Local validation:

- focused logservice regression, `-count=100`: passed;
- `go test ./pkg/sql/plan/function ./pkg/sql/plan -count=1`: passed;
- `make build`: passed;
- `git diff --check`: passed.

A fresh CI run has been triggered by the push.

File: pkg/sql/plan/base_binder.go [hunk 4]
```diff
@@ -6058,6 +6038,90 @@ func bindFuncExprImplByPlanExpr(
...
+		// remains executable by older CNs during a rolling upgrade.  Only the
+		// precision and scale metadata may be refined here.
+		if digits >= int64(input.Scale) {
+			return
+		}
+		resultScale := int32(0)
+		if digits > 0 {
+			resultScale = int32(digits)
+		}
+		precision := integerDigits + resultScale + 1 // reserve a carry digit
```
Comment (XuPeng-SH) [inline pkg/sql/plan/base_binder.go:6100]: [P2] Bound carry precision for negative-digit CEIL/FLOOR

For input DECIMAL(65,0) and literal digits=-1, this path computes integerDigits=65, resultScale=0, then Width=66 while retaining Decimal256. This differs from the former width-65 metadata even for a small, valid value. Example: `CREATE TABLE src(d DECIMAL(65,0)); INSERT INTO src VALUES (1); CREATE TABLE dst AS SELECT CEIL(d,-1) AS x FROM src;` (CEILING/FLOOR share the issue).

The resulting expression type is DECIMAL(66,0), and genAsSelectCols directly copies expr.Typ into the target ColDef. That lies outside the SQL decimal contract: mysql_sql.y rejects declared precision >65. The change can therefore expose/persist metadata that cannot be recreated as an equivalent DECIMAL declaration, even though rounding the input value itself fits easily. No downstream clamp exists in this binder helper/CTAS type derivation.

Please preserve the supported precision bound for negative D (as the ROUND/TRUNCATE branch already preserves a conservative width), or apply a defined capped precision/overflow policy without creating an unsupported type. Add maximum-width negative-digit binder and CTAS/schema-roundtrip controls; do not fix this by narrowing values to BIGINT. This is source-derived, not a claimed executed CTAS reproduction.

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
Re-reviewed the complete current four-file change at 77537dea5bb73b994a54d22b0243c5f51404f0a9 against main b349fea56f1c16e3c947ac730cc114fca3c3f241. Prior DECIMAL(19,0) wrapping, incompatible BIGINT physical lowering and missing two-argument CEIL/FLOOR refinement are fixed. One P2 maximum-precision metadata blocker remains (inline).

The revised binder preserves Decimal64/128/256 OIDs, so the prior old-CN result-wrapper failure is removed. Nonnegative ROUND/TRUNCATE precision reduction distinguishes carry vs truncation and preserves dynamic/NULL fallback; boundary execution tests now include both signed-int64 neighbors and full 19-digit extrema. The ancillary port fixture reserves TCP while choosing UDP, bounds retries to EADDRINUSE and retains the failed-probe cleanup oracle; it is not evidence about rounding semantics.

Evidence: full diff/history and prior inline findings, exact-head binder/overload/evaluator and CTAS/parser consumers inspected. Reused the author's corrected package/build report and scoped repeated port test report; the earlier malformed comment containing failed build output is not treated as pass evidence. I did not run SQL/native tests, CTAS restore, mixed-version deployment or benchmarks; the finding is source-derived. Head/base rechecked before posting; CI not awaited.

File: pkg/sql/plan/base_binder.go (hunk 4 above)
Comment (daviszhen) [reply pkg/sql/plan/base_binder.go:6100]: Implemented in cdfb676030. Negative-digit CEIL, CEILING, and FLOOR now cap result precision to the physical decimal family limit (18, 38, or 65) while preserving the decimal type. Added maximum-width binder cases and a CTAS/schema round-trip test for DECIMAL(65,0). Synced current main at bfed62b4b1; focused plan tests and make build pass.

**Review by XuPeng-SH — APPROVED** [review]
Re-reviewed bfed62b4b1b10b012b07320161f4df77e0fdf747 against main@77dbf5464466b1d5ae784e7d17d92cd1c1c8d17d. No remaining concrete blocker found; prior requested changes are closed.

The DECIMAL(65,0), digits=-1 CEIL/CEILING/FLOOR counterexample now caps precision at 65 rather than publishing width 66. The new helper also respects Decimal64/128 bounds while retaining their physical OIDs. Added tests cover family maxima, carry below the maximum, and CTAS plus an equivalent schema parse/build round-trip. Prior DECIMAL(19,0) wrapping and old-CN incompatible BIGINT lowering remain removed; two-argument refinement remains present. Nonnegative ROUND carry precision cannot exceed the input width because that branch requires digits<input.Scale; TRUNCATE excludes the carry. Dynamic/NULL and negative ROUND/TRUNCATE fallbacks are unchanged.

Reviewed the full five-file PR diff, prior inline findings/replies and current delta, including the separate port-fixture fix. No new executor, wire, persisted representation or shared lifecycle is introduced by the rounding change; ordinary-fix design exemption applies. The port helper bounds retry to address-in-use, releases its TCP reservation, and preserves the failed-probe cleanup oracle.

Evidence: reused source review and author-reported focused plan/package/build validation; the new CTAS test exercises schema expressibility, not an executed backup/restore deployment. No independent native suite, mixed-version deployment, benchmark or CI wait. Exact live head/base rechecked immediately before submission.

**Comment by daviszhen** [comment]
Synced with the latest upstream main (6602ca0ffa).\n\nLocal verification:\n- go test ./pkg/sql/plan ./pkg/sql/plan/function -count=1\n- make build\n\nAll passed. The branch is no longer behind main and GitHub reports it as mergeable. New head: 03d97de3bfbabdf7ec6a159b2fbe051046c6b1be.

**Comment by daviszhen** [comment]
Resolved the Ubuntu/x86 UT failure at head `f0c708ef8d`.

Root cause from `ut-diagnostics-34755937428-1-all`: `TestISCPResumeRecoversAcceptedIteration` resumed the executor at 12:55:02, but the recovery path's consumer-table DDL did not commit until 12:55:16 under CI load. The test's 10-second `Eventually` deadline expired at 12:55:12 and test cleanup then stopped the executor. Therefore the subsequent `context canceled` / `update txn snapshot` messages were cleanup consequences, not the initiating failure.

Fix:
- give the recovery completion assertion 30 seconds of headroom for its DDL and replay transaction;
- retain the test's existing 5-minute overall context deadline and 10 ms polling interval.

Local verification:
- `make build`: passed;
- `go test ./pkg/vm/engine/test -run '^TestISCPResumeRecoversAcceptedIteration$' -count=3 -timeout=5m`: passed;
- `git diff --check`: passed.

A fresh CI run has been triggered by the push.

**Review by aptend — APPROVED** [review]
Deep re-review of exact head `fd448b08353703331e1fe721fa53f36715cb09b0` against exact base `a68a200c0bfb0c18f1b7f49d26b20b8d0d9b017f`: APPROVE.

I read the complete review, inline-comment, author-reply, issue, and resolved-thread history; audited the current five-file diff and the delta since my prior reviewed head `e98c59b344e2cd7fce8b4d6eef816c9e40dde335`; and traced binding through overload selection, serialized result types, Decimal64/128/256 evaluators, result-vector scaling, CTAS/schema derivation, and the maximum-precision boundary.

All prior blockers are closed. DECIMAL(19,0) and the other decimal domains keep their existing physical OID, so values outside signed BIGINT do not wrap and old CN executors still receive the wrapper family they understand. Constant two-argument CEIL/CEILING/FLOOR refinement is restored. Carry-aware widths stay within the Decimal64/128/256 family maxima, including DECIMAL(65,0) with negative digits, and the added schema round trip prevents publishing DECIMAL(66,0). ROUND and TRUNCATE retain the conservative dynamic/NULL/negative-digits behavior while narrowing the unambiguous nonnegative literal case. The unrelated port-fixture correction and the bounded ISCP recovery deadline change are safe and scoped to tests.

Validation on the exact head: required third-party and CGO artifacts built; focused planner, function, and logservice tests passed; the changed ISCP recovery test passed; `git diff --check` and gofmt checks passed. I also added a temporary exact-head executor test for ROUND/TRUNCATE/CEIL/FLOOR over `DECIMAL(10,4)` at the carry boundary `999999.9999`, using the newly narrowed result types; all four results and result-vector types passed, and the temporary test was removed. All exact-head CI checks are terminal with no failures.
